src/encoder.ts

```diff
@@ -0,0 +1,248 @@
+	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9, 0xbdbdf21c, 0xcabac28a, 0x53b39330,
+	0x24b4a3a6, 0xbad03605, 0xcdd70693, 0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d,
+] as const;
+
+function crc32(crc: number, data: Uint8Array) {
+	crc = ~~crc ^ -1;
+	for (let n = 0; n < data.length; n++) {
+		crc = CrcTable[(crc ^ data[n]) & 0xff] ^ (crc >>> 8);
+	}
+	return crc ^ -1;
+}
+
+export class Encoder {
+	#format: Format;
+	#sampleRate: Encoding["sampleRate"];
+	#symbolDurations: Encoding["symbolDurations"];
+	#packetSize: Encoding["packetSize"];
+	#polarity = 0;
+	#samples: Float32Array[][] = [];
+
+	constructor(encoding: Encoding, format: Format) {
+		this.#sampleRate = encoding.sampleRate;
+		this.#symbolDurations = encoding.symbolDurations;
+		this.#packetSize = encoding.packetSize;
+
+		const fn = encoding.shape == "cosine" ? Math.cos : Math.sin;
+		const noDC = encoding.shape == "sine_no_dc";
+
+		for (let duration of encoding.symbolDurations) {
+			const gain = noDC ? Math.max(encoding.symbolDurations[0] / duration, 0.5) : 1.0;
+			const plus = new Float32Array(duration);
+			const minus = new Float32Array(duration);
+			for (let i = 0; i < duration; i++) {
+				plus[i] = fn((i / duration) * Math.PI) * gain;
+				minus[i] = -plus[i];
+			}
+			this.#samples.push([plus, minus]);
+		}
+		this.#format = format;
+	}
+
+	encode(symbols: Uint8Array) {
+		let totalDuration = 0;
+		for (const symbol of symbols) {
+			totalDuration += this.#symbolDurations[symbol];
+		}
+		const block = new Float32Array(totalDuration);
+		let i = 0;
+		for (const symbol of symbols) {
+			block.set(this.#samples[symbol][this.#polarity], i);
+			i += this.#symbolDurations[symbol];
+			this.#polarity = 1 - this.#polarity;
+		}
+		return block;
+	}
+
+	codeBlank(duration: number) {
+		const n = Math.ceil((duration * this.#sampleRate) / this.#symbolDurations[2]);
+		return this.encode(new Uint8Array(Array(n).fill(2)));
+	}
+
+	pad(data: Uint8Array, size: number) {
+		const n = data.length;
+		if (n % size != 0) {
+			const paddedData = new Uint8Array(n + size - (n % size));
+			paddedData.set(data);
+			data = paddedData;
+		}
+		return data;
+	}
+
+	codePacket(packet: Uint8Array) {
+		console.assert(packet.length <= this.#packetSize);
+		packet = this.pad(packet, this.#packetSize);
+		const crc = crc32(0, packet);
+		const header = new Uint8Array([0x55, 0x55, 0x55, 0x55]);
+		const footer = new Uint8Array([crc >>> 24, (crc >>> 16) & 0xff, (crc >>> 8) & 0xff, crc & 0xff]);
+		const symbolStream = [];
+		for (const block of [header, packet, footer]) {
+			for (const byte of block) {
+				for (let bit = 0; bit < 8; bit++) {
+					symbolStream.push(byte & (0x80 >> bit) ? 1 : 0);
+				}
+			}
+		}
+		return this.encode(new Uint8Array(symbolStream));
+	}
+
+	code(data: Uint8Array, tags: number[]) {
```

Comment:
@pichenettes I'm pretty this method take in a `Uint8Array`s but I would appreciate it if you reviewed the method parameter types for this class. I think the refactor of `Encoder` is complete.